compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+                   throwIfDifferent "observers" (extractFuncFromFuncThisArg <$> Upgrading past present) observersErr
+               _ -> do
+                   throwWithContext observersErr
+
+            let authorizersErr = EUpgradeChoiceChangedAuthorizers $ NM.name $ present choice
+            case fmap (mapENilToNothing . chcAuthorizers) choice of
+               Upgrading { past = Nothing, present = Nothing } -> pure ()
+               Upgrading { past = Just past, present = Just present } ->
+                   throwIfDifferent "authorizers" (extractFuncFromFuncThisArg <$> Upgrading past present) authorizersErr
+               _ -> throwWithContext authorizersErr
+        pure choice
+
+    -- This check assumes that we encode signatories etc. on a template as
+    -- $<uniquename> this, where $<uniquename> is a function that contains the
+    -- actual definition. We resolve this function and check that it is
+    -- identical.
+    withContext (ContextTemplate (present module_) (present template) TPPrecondition) $
+        throwIfDifferent "precondition" (extractFuncFromCaseFuncThis . tplPrecondition <$> template) $
+            EUpgradeTemplateChangedPrecondition $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPSignatories) $
+        throwIfDifferent "signatories" (extractFuncFromFuncThis . tplSignatories <$> template) $
+            EUpgradeTemplateChangedSignatories $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPObservers) $
+        throwIfDifferent "observers" (extractFuncFromFuncThis . tplObservers <$> template) $
+            EUpgradeTemplateChangedObservers $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPAgreement) $
+        throwIfDifferent "agreement" (extractFuncFromFuncThis . tplAgreement <$> template) $
+            EUpgradeTemplateChangedAgreement $ NM.name $ present template
+    -- TODO: Check that return type of a choice is compatible
+    pure ()
+    where
+        extractFuncFromFuncThis expr
+            | ETmApp{..} <- expr
+            , EVal qualEvn <- tmappFun
+            , EVar (ExprVarName "this") <- tmappArg
+            = Just (qualObject qualEvn)
+            | otherwise
+            = Nothing
+        extractFuncFromFuncThisArg expr
+            | outer@ETmApp{} <- expr
+            , EVar (ExprVarName "arg") <- tmappArg outer
+            , inner@ETmApp{} <- tmappFun outer
+            , EVar (ExprVarName "this") <- tmappArg inner
+            , EVal qualEvn <- tmappFun inner
+            = Just (qualObject qualEvn)
+            | otherwise
+            = Nothing
+        extractFuncFromCaseFuncThis expr
+            | ECase{..} <- expr
+            = extractFuncFromFuncThis casScrutinee
+            | otherwise
+            = Nothing
+        resolveExpression field expr module_ =
```

Comment:
Some of these might benefit from type sigs

Reply:
Fixed